edb/server/protocol/execute.pyx

```diff
@@ -585,3 +586,64 @@ cdef bytes _encode_args(list args):
                 out_buf.write_bytes(jval)
 
     return bytes(out_buf)
+
+
+async def interpret_error(
+    exc: Exception,
+    db: dbview.Database,
+    *,
+    global_schema_pickle: object=None,
+    user_schema_pickle: object=None,
+    from_graphql: bool=False,
+) -> tuple[Exception, type]:
+
+    if isinstance(exc, RecursionError):
+        exc = errors.UnsupportedFeatureError(
+            "query too deeply nested: the query caused the compiler "
+            "stack to overflow",
+        )
+
+    elif isinstance(exc, pgerror.BackendError):
+        try:
+            static_exc = errormech.static_interpret_backend_error(
+                exc.fields, from_graphql=from_graphql
+            )
+
+            # only use the backend if schema is required
+            if static_exc is errormech.SchemaRequired:
+                user_schema_pickle = (
+                    user_schema_pickle or db.user_schema_pickle)
+                global_schema_pickle = (
+                    global_schema_pickle or db._index._global_schema_pickle
+                )
+                compiler_pool = db._index._server.get_compiler_pool()
+                exc = await compiler_pool.interpret_backend_error(
+                    user_schema_pickle,
+                    global_schema_pickle,
+                    exc.fields,
+                    from_graphql,
+                )
+
+            elif isinstance(static_exc, (
+                    errors.DuplicateDatabaseDefinitionError,
+                    errors.UnknownDatabaseError)):
+                tenant_id = db.tenant.tenant_id
+                message = static_exc.args[0].replace(f'{tenant_id}_', '')
+                exc = type(static_exc)(message)
+            else:
+                exc = static_exc
+
+        except Exception as e:
+            from edb.common import debug
+            if debug.flags.server:
+                debug.dump(e)
+
+            exc = RuntimeError(
+                'unhandled error while calling interpret_backend_error(); '
+                'run with EDGEDB_DEBUG_SERVER to debug.')
+
+    exc_type = type(exc)
+    if not issubclass(exc_type, errors.EdgeDBError):
+        exc_type = errors.InternalServerError
+
+    return exc, exc_type
```

Comment:
Can you elaborate a bit please why this instead of chaining or wrapping?

Reply:
Let me try changing it

Reply:
Changed